<a href="https://colab.research.google.com/github/jsvitorm/repositorioGITTEST/blob/master/Sprint2_de_Grupo2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

IMPORTAÇÃO DE BIBLIOTECAS E DATAFRAMES

Nessa cédula de código temos a importação das bases de dados. Estamos utilizando 2 bases de dados fornecidas pelo cliente: "Acompanhamento" e "Fornecedores".

A base de dados "Acompanhamento" será nossa base de dados principal e a "Fornecedores" será utilizada apenas para integrarmos a coluna "CNPJ" que está presente apenas na base de dados "Fornecedores".


In [ ]:
#DOWNLOADS

!pip install scikit-surprise

In [ ]:
import pandas as pd
import plotly.express as px
import matplotlib.pyplot as plt
import gdown
import numpy as np
from google.colab import files
from surprise import SVD, Reader
from surprise import Dataset, accuracy
from surprise.model_selection import cross_validate, train_test_split
import joblib


# Nome do arquivo destino no Colab
arquivo_destino_no_colab = "dataset.csv"
arquivo_destino_no_colab2 = "dataset_fornecedores.csv"

doc_id = "1ohTZvdq08VPhOrILLZOHQstNFYyMAf5A"
doc_id2 = "1aJUFr2oOrA3NRrcsgfqJHNxGKrKg1_vG"

# Carrega o dataset usando gdown
URL = f"https://drive.google.com/uc?id={doc_id}"

gdown.download(URL, arquivo_destino_no_colab, quiet=False)

URL2 = f"https://drive.google.com/uc?id={doc_id2}"

gdown.download(URL2, arquivo_destino_no_colab2, quiet=False)


forn = pd.read_csv(arquivo_destino_no_colab2, on_bad_lines='skip', sep='\t')
dados = pd.read_csv(arquivo_destino_no_colab)

display(dados)


Downloading...
From: https://drive.google.com/uc?id=1ohTZvdq08VPhOrILLZOHQstNFYyMAf5A
To: /content/dataset.csv
100%|██████████| 110M/110M [00:00<00:00, 284MB/s] 
Downloading...
From: https://drive.google.com/uc?id=1aJUFr2oOrA3NRrcsgfqJHNxGKrKg1_vG
To: /content/dataset_fornecedores.csv
100%|██████████| 28.4k/28.4k [00:00<00:00, 46.5MB/s]
<ipython-input-2-e880c681f5aa>:32: DtypeWarning: Columns (10,12) have mixed types. Specify dtype option on import or set low_memory=False.
  dados = pd.read_csv(arquivo_destino_no_colab)


Contrato     GR/GPB  \
0                                36/00817/16/05-001  1420662.0   
1                               36/00235/17/05-001B  1430636.0   
2                               36/00235/17/05-001B  1430636.0   
3                               36/00235/17/05-001B  1430637.0   
4                                36/00854/18/05-001  1447250.0   
...                                             ...        ...   
465166  MULTI - NOTEBOOK SALA DE AULA - 2022NE00724  1702820.0   
465167   MULTILASER - Notebook Básico - 2020NE00274  1648358.0   
465168                                  57/00010/21  1660963.0   
465169   MULTILASER - Notebook Básico - 2020NE00274  1648359.0   
465170    MULTILASER - TABLET ANDROID - 2020NE00264  1510940.0   

       Data do Contrato Emissão GPB Data da Baixa Baixa  Qtde.  \
0            01/11/2016  03/11/2016    15/02/2022   Sim    2.0   
1            02/06/2017  01/07/2017    15/02/2022   Sim    2.0   
2            02/06/2017  01/07/2017    15/02/2022   Sim    1.0   
3            02/06/2017  01/07/2017    01/08/2017   Sim    2.0   
4            19/10/2018  22/10/2018           Não   Não   30.0   
...                 ...         ...           ...   ...    ...   
465166       30/12/2022  24/08/2023           Não   Não  120.0   
465167       28/10/2020  20/12/2022           Não   Não   15.0   
465168       29/12/2022  20/01/2023    14/08/2023   Sim   10.0   
465169       28/10/2020  20/12/2022    14/08/2023   Sim   30.0   
465170       30/12/2020  19/05/2021    13/07/2021   Sim    9.0   

                                       Material  \
0       CAIXA PARA ACONDICIONAMENTO DE LÂMPADAS   
1       CAIXA PARA ACONDICIONAMENTO DE LÂMPADAS   
2       CAIXA PARA ACONDICIONAMENTO DE LÂMPADAS   
3       CAIXA PARA ACONDICIONAMENTO DE LÂMPADAS   
4       CAIXA PARA ACONDICIONAMENTO DE LÂMPADAS   
...                                         ...   
465166              NOTEBOOK SALA DE AULA MULTI   
465167              Notebook Básico Educacional   
465168                         APARELHO CELULAR   
465169              Notebook Básico Educacional   
465170                          Tablets Android   

                                                     Item    Diretoria    CIE  \
0       Tipo 1 - Caixa para acondicionamento de lâmpad...   ADAMANTINA  31264   
1       Tipo 1 - Caixa para acondicionamento de lâmpad...   ADAMANTINA  31264   
2       Tipo 2 - Caixa para acondicionamento de lâmpad...   ADAMANTINA  31264   
3       Tipo 1 - Caixa para acondicionamento de lâmpad...   ADAMANTINA  20901   
4       Tipo 1 - Caixa para acondicionamento de lâmpad...   ADAMANTINA  20901   
...                                                   ...          ...    ...   
465166                        NOTEBOOK SALA DE AULA MULTI  VOTUPORANGA  29208   
465167                                              UL150  VOTUPORANGA  29208   
465168                                      MULTILASER G2  VOTUPORANGA  20711   
465169                                              UL150  VOTUPORANGA  20711   
465170                                     Tablet Android  VOTUPORANGA  20711   

                             Escola Data de entrega  \
0                        9 DE JULHO             NaN   
1                        9 DE JULHO             NaN   
2                        9 DE JULHO             NaN   
3                        ADAMANTINA             NaN   
4                        ADAMANTINA             NaN   
...                             ...             ...   
465166  UZENIR COELHO ZEITUNE PROFA      24/08/2023   
465167  UZENIR COELHO ZEITUNE PROFA      20/12/2022   
465168                  VOTUPORANGA      23/01/2023   
465169                  VOTUPORANGA      20/12/2022   
465170                  VOTUPORANGA      19/05/2021   

                              Classe do Material  \
0                    Outros Materiais de Consumo   
1                    Outros Materiais de Consumo   
2                    Outros Materiais de Consumo   
3 

Nessa cédula de código estamos filtrando apenas as colunas "ESCOLA", "CIE" "FORNECEDOR" e "AVALIAÇÃO" para nosso dataframe final. De 16 colunas do dataframe original, estamos utilizando apenas 4 colunas para essa análise introdutória.

Possuímos 465171 linhas, essas linhas representam uma entrega: A escola destinatária, o código único da escola (CIE), a fornecedora responsável e a avaliação da escola em relação a entrega.

In [ ]:
df_dados = dados


df_dados = df_dados[['Escola','CIE','Fornecedor','Avaliação']]

df_temp = df_dados

display(df_dados)

Escola    CIE  \
0                        9 DE JULHO  31264   
1                        9 DE JULHO  31264   
2                        9 DE JULHO  31264   
3                        ADAMANTINA  20901   
4                        ADAMANTINA  20901   
...                             ...    ...   
465166  UZENIR COELHO ZEITUNE PROFA  29208   
465167  UZENIR COELHO ZEITUNE PROFA  29208   
465168                  VOTUPORANGA  20711   
465169                  VOTUPORANGA  20711   
465170                  VOTUPORANGA  20711   

                                               Fornecedor  Avaliação  
0       PROVISÃO JC COMÉRCIO DE ELETRO ELETRÔNICOS LTDA -          3  
1       INDÚSTRIA E COMÉRCIO ELETRO ELETRÔNICA GEHAKA ...          0  
2       INDÚSTRIA E COMÉRCIO ELETRO ELETRÔNICA GEHAKA ...          4  
3       PROVISÃO JC COMÉRCIO DE ELETRO ELETRÔNICOS LTDA -          5  
4       PROVISÃO JC COMÉRCIO DE ELETRO ELETRÔNICOS LTDA -          4  
...                                                   ...        ...  
465166                    AGEM TECNOLOGIA E COMERCIO LTDA          4  
465167                           POSITIVO TECNOLOGIA S.A.          0  
465168  SEATTLE TECNOLOGIA E COMÉRCIO DE PRODUTOS ELET...          5  
465169                         POSITIVO INFORMÁTICA S. A.          3  
465170                   LENOVO TECNOLOGIA (BRASIL) LTDA.          3  

[465171 rows x 4 columns]

Aqui adicionamos a coluna "CNPJ" ao nosso dataframe. Agora possuímos 5 colunas. Esse CNPJ é referente à Fornecedora.

In [ ]:
df_completo = df_temp.merge(forn, left_on='Fornecedor', right_on='RazaoSocial', how='left')

df_dados['CNPJ'] = df_completo['CNPJ']

df_dadosneed = df_dados[["CIE", "CNPJ", "Avaliação"]]

df_dados


<ipython-input-4-2fc3e77caa15>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_dados['CNPJ'] = df_completo['CNPJ']


Escola    CIE  \
0                        9 DE JULHO  31264   
1                        9 DE JULHO  31264   
2                        9 DE JULHO  31264   
3                        ADAMANTINA  20901   
4                        ADAMANTINA  20901   
...                             ...    ...   
465166  UZENIR COELHO ZEITUNE PROFA  29208   
465167  UZENIR COELHO ZEITUNE PROFA  29208   
465168                  VOTUPORANGA  20711   
465169                  VOTUPORANGA  20711   
465170                  VOTUPORANGA  20711   

                                               Fornecedor  Avaliação  \
0       PROVISÃO JC COMÉRCIO DE ELETRO ELETRÔNICOS LTDA -          3   
1       INDÚSTRIA E COMÉRCIO ELETRO ELETRÔNICA GEHAKA ...          0   
2       INDÚSTRIA E COMÉRCIO ELETRO ELETRÔNICA GEHAKA ...          4   
3       PROVISÃO JC COMÉRCIO DE ELETRO ELETRÔNICOS LTDA -          5   
4       PROVISÃO JC COMÉRCIO DE ELETRO ELETRÔNICOS LTDA -          4   
...                                                   ...        ...   
465166                    AGEM TECNOLOGIA E COMERCIO LTDA          4   
465167                           POSITIVO TECNOLOGIA S.A.          0   
465168  SEATTLE TECNOLOGIA E COMÉRCIO DE PRODUTOS ELET...          5   
465169                         POSITIVO INFORMÁTICA S. A.          3   
465170                   LENOVO TECNOLOGIA (BRASIL) LTDA.          3   

                      CNPJ  
0       10.284.501/0001-09  
1       60.637.667/0001-21  
2       60.637.667/0001-21  
3       10.284.501/0001-09  
4       10.284.501/0001-09  
...                    ...  
465166  09.022.398/0001-31  
465167  81.243.735/0019-77  
465168  23.556.435/0001-12  
465169  81.243.735/0001-48  
465170  07.275.920/0001-61  

[465171 rows x 5 columns]

Aqui estamos fazendo uma breve análise para entender o contexto das avaliações das entregas. Podemos analisar que a maioria das notas se concentram em "1", "4" e "5". Isso nos dá a entender que as avaliações tendem a ser mais "extremistas".

In [ ]:
df_avaliacao_por_nota = df_dados['Avaliação'].value_counts().reset_index()

px.bar(df_avaliacao_por_nota, x='index', y='Avaliação', title="Avaliações de Entregas")

Aqui estamos analisando quantos fornecedores e quantas escolas temos disponibilizados na base de dados.

In [ ]:
print(df_dados['Fornecedor'].nunique())
print(df_dados['Escola'].nunique())



101
5886


Nesse trecho, estamos analisando quantas avaliações cada escola fez e a média das avaliações que essa escola fornece.
Além disso, filtramos para pegar dados apenas de escolas que fizeram pelo menos 8 avaliações para evitar que a média de avaliações da escola seja comprometida por conta da baixa quantidade de avaliações. Esse filtro ajuda a diminuir a quantidade de outliers


In [ ]:
media_ava = df_dados['Avaliação'].mean()
media_ava

escolas = df_dados.groupby('Escola')['Avaliação'].agg(['mean', 'count'])
escolas.columns = ['Média', 'Quantidade de Avaliações']

# Filtra escolas com quantidade de avaliações acima de 8
escolas_filtradas = escolas[escolas['Quantidade de Avaliações'] > 8]

# Filtra o DataFrame original mantendo apenas as escolas com mais de 8 avaliações
df_dados = df_dados[df_dados['Escola'].isin(escolas_filtradas.index)]

# Ordena o DataFrame resultante pela quantidade de avaliações em ordem decrescente
#df_dados.sort_values(by='Quantidade de Avaliações', ascending=False, inplace=True)

# Exibe o DataFrame resultante
print(escolas_filtradas)


                                          Média  Quantidade de Avaliações
Escola                                                                   
16 DE JULHO                            2.925000                        80
17 DE SETEMBRO                         3.032609                        92
18 DE JUNHO                            2.618056                       144
20 DE AGOSTO                           2.662651                        83
21 DE ABRIL                            2.616279                        86
...                                         ...                       ...
ZULMIRA CAVALHEIRO FAUSTINO DONA       2.964286                        84
ZULMIRA DA SILVA SALLES PROFESSORA     2.695122                        82
ZULMIRA DE ALMEIDA LAMBERT PROFESSORA  2.645833                       144
ZULMIRA DE OLIVEIRA PROFESSORA         2.811881                       101
ZULMIRO ALVES DE SIQUEIRA              2.988235                        85

[5401 rows x 2 columns]


Aqui desenvolvemos um gráfico para tornar a informação anterior mais visual. Podemos observar que nosso X é a quantidade de avaliações e nosso Y é a nota média das avaliações que aquela escola fornece. Quanto mais pra cima o ponto se encontra, significa que aquela escola tende a avaliar mais positivamente as fornecedoras. Quanto mais para a direita o ponto se encontrar, significa que aquela escola avalia mais vezes.

In [ ]:
fig = px.scatter(escolas_filtradas, x="Quantidade de Avaliações", y="Média")
fig.show()

Aqui estamos fazendo o mesmo procedimento que fizemos para as escolas, porém para as fornecedoras. Nesse trecho em específico, estamos analisando a quantidade de avaliações que a fornecedora recebeu, e a nota média que ela recebe nessas avaliações.

In [ ]:

fornecedores = df_dados.groupby('Fornecedor')['Avaliação'].agg(['mean', 'count'])
fornecedores.columns = ['Média', 'Quantidade de Avaliações']

# Filtra fornecedores com média acima de 2
fornecedores_filtrados = fornecedores[fornecedores['Média'] > 2]

# Filtra o DataFrame original mantendo apenas os fornecedores com média acima de 2
df_dados = df_dados[df_dados['Fornecedor'].isin(fornecedores_filtrados.index)]

# Ordena o DataFrame resultante pela média em ordem crescente
df_dados.sort_values(by='Avaliação', ascending=True, inplace=True)

# Exibe o DataFrame resultante
display(df_dados)


<ipython-input-9-4f43cb6d9363>:11: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



Escola       CIE  \
240699                   NARCISO YAGUE GUIMARAES VEREADOR    041130   
333588  MARIA APARECIDA FRANCA BARBOSA DE ARAUJO PROFE...  906748.0   
333591                     MARIO CARDOSO FRANCO PROFESSOR   14285.0   
333596                           PEREIRA DE MATTOS DOUTOR   49219.0   
106486                     OACYR ANTONIO ELLERO PROFESSOR    462536   
...                                                   ...       ...   
281231              SANDRA APARECIDA DE ARAUJO PROFESSORA    924337   
158593                    DALVA LELLIS GARCIA PRADO PROFA   22421.0   
78020                                  OSWALDO SAMMARTINO     10662   
366707                           ELOI LOPES FERRAZ DOUTOR   28125.0   
350510                                            LIMEIRA   20408.0   

                                               Fornecedor  Avaliação  \
240699                  RJI CURSOS E CONCURSOS LTDA - EPP          0   
333588                     ESDEVA - INDUSTRIA GRÁFICA S/A          0   
333591  RIO OFFICE COMÉRCIO DE MOVEIS E EQUIPAMENTOS E...          0   
333596  WEBLABOR SÃO PAULO MATERIAIS DIDÁTICOS LTDA - EPP          0   
106486                    AGEM TECNOLOGIA E COMERCIO LTDA          0   
...                                                   ...        ...   
281231                  RJI CURSOS E CONCURSOS LTDA - EPP          5   
158593                       TSW COMÉRCIO DE MÓVEIS LTDA.          5   
78020   PROVISÃO JC COMÉRCIO DE ELETRO ELETRÔNICOS LTDA -          5   
366707          INDÚSTRIA E COMÉRCIO DE MÓVEIS NAÇÃO LTDA          5   
350510                   LENOVO TECNOLOGIA (BRASIL) LTDA.          5   

                      CNPJ  
240699  15.328.920/0001-00  
333588  17.153.081/0001-62  
333591  11.496.190/0001-04  
333596  13.533.610/0001-00  
106486  09.022.398/0001-31  
...                    ...  
281231  15.328.920/0001-00  
158593  10.456.190/0001-00  
78020   10.284.501/0001-09  
366707  10.841.510/0001-45  
350510  07.275.920/0001-61  

[318086 rows x 5 columns]

Nesse gráfico de [Quantidade de Avaliações x Média] podemos observar que não há fornecedoras com nota média entre 2 e 2,8.

In [ ]:
fig = px.scatter(fornecedores_filtrados, x="Quantidade de Avaliações", y="Média")
fig.show()

Nessa parte, desenvolvemos uma matriz usuário-item, considerando que a escola como usuário e os fornecedores como itens. Caso uma escola ainda não tenha avaliado uma fornecedora essa célula será preenchida com a média do fornecedor correspondente

In [ ]:
media_fornecedores = df_dados.groupby('Fornecedor')['Avaliação'].mean()


matriz_usuario_item = pd.pivot_table(df_dados, values='Avaliação', index='Escola', columns='Fornecedor', aggfunc='mean')


for fornecedor, media in media_fornecedores.items():
    matriz_usuario_item[fornecedor].fillna(media, inplace=True)


display(matriz_usuario_item)

matriz_usuario_item.to_csv('matriz_usuario_item.csv')

#files.download('matriz_usuario_item.csv')

Fornecedor                             ADTK ATACADO IMPORTAÇÃO E EXPORTAÇÃO DE PRODUTOS E  \
Escola                                                                                      
16 DE JULHO                                                                     2.939735    
17 DE SETEMBRO                                                                  2.939735    
18 DE JUNHO                                                                     2.939735    
20 DE AGOSTO                                                                    2.939735    
21 DE ABRIL                                                                     2.939735    
...                                                                                  ...    
ZULMIRA CAVALHEIRO FAUSTINO DONA                                                5.000000    
ZULMIRA DA SILVA SALLES PROFESSORA                                              2.939735    
ZULMIRA DE ALMEIDA LAMBERT PROFESSORA                                           2.939735    
ZULMIRA DE OLIVEIRA PROFESSORA                                                  2.939735    
ZULMIRO ALVES DE SIQUEIRA                                                       2.939735    

Fornecedor                             AGEM TECNOLOGIA E COMERCIO LTDA  \
Escola                                                                   
16 DE JULHO                                                   4.000000   
17 DE SETEMBRO                                                4.000000   
18 DE JUNHO                                                   2.960145   
20 DE AGOSTO                                                  4.000000   
21 DE ABRIL                                                   3.000000   
...                                                                ...   
ZULMIRA CAVALHEIRO FAUSTINO DONA                              3.500000   
ZULMIRA DA SILVA SALLES PROFESSORA                            2.960145   
ZULMIRA DE ALMEIDA LAMBERT PROFESSORA                         3.333333   
ZULMIRA DE OLIVEIRA PROFESSORA                                2.960145   
ZULMIRO ALVES DE SIQUEIRA                                     3.000000   

Fornecedor                             AGUAMAR TRANSPORTES LTDA  \
Escola                                                            
16 DE JULHO                                            2.984794   
17 DE SETEMBRO                                         2.984794   
18 DE JUNHO                                            4.500000   
20 DE AGOSTO                                           3.000000   
21 DE ABRIL                                            2.984794   
...                                                         ...   
ZULMIRA CAVALHEIRO FAUSTINO DONA                       1.000000   
ZULMIRA DA SILVA SALLES PROFESSORA                     1.000000   
ZULMIRA DE ALMEIDA LAMBERT PROFESSORA                  2.984794   
ZULMIRA DE OLIVEIRA PROFESSORA                         3.000000   
ZULMIRO ALVES DE SIQUEIRA                              2.984794   

Fornecedor                             AIDC TECNOLOGIA LTDA  \
Escola                                                        
16 DE JULHO                                        2.972965   
17 DE SETEMBRO                                     3.000000   
18 DE JUNHO                                        2.972965   
20 DE AGOSTO                                       2.972965   
21 DE ABRIL                                        2.972965   
...                                                     ...   
ZULMIRA CAVALHEIRO FAUSTINO DONA                   2.972965   
ZULMIRA DA SILVA SALLES PROFESSORA                 5.000000   
ZULMIRA DE ALMEIDA LAMBERT PROFESSORA              3.000000   
ZULMIRA DE OLIVEIRA PROFESSORA                     5.000000   
ZULMIRO ALVES DE SIQUEIRA                          2.972965   

Fornecedor                             ALLERBEST COMERCIO DE PRODUTOS PARA LABORATORIO LT  \
Escola                                    

Aqui estamos pegando informações gerais referentes a nossa matriz usuário-item, Podemos analisar a média, nota minima, nota máxima, desvio padrão e etc.

In [ ]:
matriz_usuario_item.describe()

Fornecedor  ADTK ATACADO IMPORTAÇÃO E EXPORTAÇÃO DE PRODUTOS E  \
count                                             5394.000000    
mean                                                 2.936166    
std                                                  0.850189    
min                                                  0.000000    
25%                                                  2.939735    
50%                                                  2.939735    
75%                                                  2.939735    
max                                                  5.000000    

Fornecedor  AGEM TECNOLOGIA E COMERCIO LTDA  AGUAMAR TRANSPORTES LTDA  \
count                           5394.000000               5394.000000   
mean                               2.964140                  2.991298   
std                                1.200760                  1.059641   
min                                0.000000                  0.000000   
25%                                2.960145                  2.984794   
50%                                2.960145                  2.984794   
75%                                3.000000                  3.000000   
max                                5.000000                  5.000000   

Fornecedor  AIDC TECNOLOGIA LTDA  \
count                5394.000000   
mean                    2.972170   
std                     1.174839   
min                     0.000000   
25%                     2.972965   
50%                     2.972965   
75%                     3.000000   
max                     5.000000   

Fornecedor  ALLERBEST COMERCIO DE PRODUTOS PARA LABORATORIO LT  \
count                                             5394.000000    
mean                                                 2.981457    
std                                                  0.357189    
min                                                  0.000000    
25%                                                  2.982906    
50%                                                  2.982906    
75%                                                  2.982906    
max                                                  5.000000    

Fornecedor  AMBY SERVICE LTDA  \
count             5394.000000   
mean                 2.835149   
std                  0.652203   
min                  0.000000   
25%                  2.832117   
50%                  2.832117   
75%                  2.832117   
max                  5.000000   

Fornecedor  APLIQUIM EQUIPAMENTOS E PRODUTOS QUIMICOS LTDA-EPP  \
count                                             5394.000000    
mean                                                 3.018401    
std                                                  0.353780    
min                                                  0.000000    
25%                                                  3.020243    
50%                                                  3.020243    
75%                                                  3.020243    
max                                                  5.000000    

Fornecedor  ARAGUAIA INDÚSTRIA GRÁFICA EDITORA  ARMAZENA ARMAZENS GERAIS LTDA  \
count                              5394.000000                    5394.000000   
mean                                  2.922171                       2.907785   
std                                   1.083294                       1.153324   
min                                   0.000000                       0.000000   
25%                                   2.932578                       2.923283   
50%                                   2.932578                       2.923283   
75%                                   3.000000                       3.000000   
max                                   5.000000                       5.000000   

Fornecedor  ARMAZENA INDÚSTRIA DE MÓVEIS LTDA EPP  ...  \
count                                 5394.000000  ...   
mean                                     2.956895  ...   
std                           

Podemos perceber que não há nenhuma célula vazia

In [ ]:
display(matriz_usuario_item.isnull().sum())

Fornecedor
ADTK ATACADO IMPORTAÇÃO E EXPORTAÇÃO DE PRODUTOS E    0
AGEM TECNOLOGIA E COMERCIO LTDA                       0
AGUAMAR TRANSPORTES LTDA                              0
AIDC TECNOLOGIA LTDA                                  0
ALLERBEST COMERCIO DE PRODUTOS PARA LABORATORIO LT    0
                                                     ..
VECTOR MOVEIS CORPORATIVOS LTDA.                      0
VENTISOL INDÚSTRIA E COMÉRCIO S.A.                    0
WEBLABOR SÃO PAULO MATERIAIS DIDÁTICOS LTDA - EPP     0
WELLINGTON SCARPARO BOTARO ME                         0
XP ON CONSULTORIA LTDA                                0
Length: 87, dtype: int64

In [ ]:

matriz_usuario_item.to_csv('MatrizGrupo2.csv',index=False)

CRIAMOS UM DF COM AS COLUNAS CIE (IDENTIFICADOR DO USUARIO ESCOLA), CNPJ (IDENTIFICADOR DO USUARIO FORNECEDOR) E AVALIAÇÃO

(OS DADOS VAZIOS ESTAMOS PREENCHENDO COM A MÉDIA DE AVALIAÇÃO DAQUELE FORNECEDOR)

ESTAMOS AGRUPANDO POR CIE E CNPJ

In [ ]:
df_dadosneed
df_agrupado = df_dadosneed.groupby(['CIE', 'CNPJ'])['Avaliação'].mean().reset_index()

In [ ]:
reader = Reader(rating_scale=(0, 5))
data = Dataset.load_from_df(df_agrupado[['CIE', 'CNPJ', 'Avaliação']], reader)

Estamos criando nosso modelo de recomendação utilizando SVD. O SVD divide essa matriz de utilidade em três outras matrizes: U (matriz de usuários), Σ (matriz diagonal de valores singulares), e V^T (a transposta da matriz de itens).

Para prever a avaliação de um usuário para um item que ele ainda não avaliou, fazemos isso multiplicando a linha correspondente do usuário (U) a matriz de valores singulares truncados (Σ) e a coluna correspondente do item (V^T).

As métricas que escolhemos foram RMSE e MAE.

RMSE (Root Mean Square Error) e MAE (Mean Absolute Error) são duas métricas usadas para medir quão bem um modelo está prevendo algo comparado aos valores reais.

O RMSE é como uma média dos erros ao quadrado, dando mais peso a erros maiores. É como calcular a raiz quadrada da média dos erros ao quadrado. Se você tiver um grande erro, ele vai afetar mais o RMSE do que o MAE.

O MAE é uma média simples dos erros absolutos, sem elevar ao quadrado. Ele trata todos os erros, grandes ou pequenos, da mesma forma.

 Bem, essas métricas ajudam a avaliar o quão preciso é o sistema ao prever preferências do usuário. Em um sistema de recomendação, você quer que as sugestões sejam o mais próximas possível das preferências reais dos usuários.
 Com essas métricas conseguimos analisar a "gravidade" dos erros do sistema de recomendação.

Também é importante mencionar que dividimos nosso dataset em 90% para treino e 10% para teste.

In [ ]:
model = SVD()

results = cross_validate(model, data, measures=['RMSE', 'MAE'], cv=3, verbose=True)


trainset, testset = train_test_split(data, test_size=0.1)

Evaluating RMSE, MAE of algorithm SVD on 3 split(s).

                  Fold 1  Fold 2  Fold 3  Mean    Std     
RMSE (testset)    1.7303  1.7196  1.7259  1.7253  0.0044  
MAE (testset)     1.4244  1.4148  1.4213  1.4202  0.0040  
Fit time          6.02    6.80    2.38    5.07    1.92    
Test time         1.26    1.11    0.67    1.01    0.25    


No geral, com base nessas métricas, parece que o modelo está performando de maneira bastante consistente em todos os folds. Embora o MAE esteja em 1.4 (relativamente alto, já que uma nota realmente 3 pode variar entre 1,6 e 4,4), o RMSE está em apenas 1.7, nos dando a entender que embora ele erre bastante, esses erros não são grosseiros.

AGORA VAMOS TESTAR UTILIZANDO NOSSO DF DE TESTES. PARA FAZER A COMPARAÇÃO, COLOCAMOS A COLUNA DE AVALIAÇÃO REAL (COMO SE FOSSE A RESPOSTA CORRETA) E LOGO EM SEGUIDA COLOCAMOS A COLUNA DE AVALIAÇÃO PREVISTA(GERADA PELO MODELO)

In [ ]:
previsoes = []

for rating in testset:
    escola = rating[0]  # CIE Escola
    fornecedor = rating[1]  # CNPJ Fornecedor
    avaliacao_real = rating[2]  # Avaliação real

    # Fazendo previsão
    previsao = model.predict(escola, fornecedor)

    avaliacao_prevista = previsao.est

    previsoes.append({'Escola': escola, 'Fornecedor': fornecedor, 'Avaliação Real': avaliacao_real, 'Avaliação Prevista': avaliacao_prevista})

#Convertendo em dataframe
df_previsoes = pd.DataFrame(previsoes)

df_previsoes

Escola          Fornecedor  Avaliação Real  Avaliação Prevista
0       12221  54.826.367/0001-98        3.333333            3.470436
1      914940  02.624.659/0001-44        5.000000            4.282720
2       29865  71.615.942/0001-22        3.000000            3.232916
3        6932  45.170.289/0001-25        4.000000            3.671504
4      922973  17.153.081/0001-62        2.750000            2.757139
...       ...                 ...             ...                 ...
21232    1855  14.516.006/0001-20        3.000000            2.719953
21233  922821  71.615.942/0001-22        4.000000            4.038461
21234   34373  10.841.510/0001-45        4.000000            2.530695
21235  297525  02.624.659/0001-44        5.000000            4.594629
21236   32585  45.170.289/0001-25        2.000000            2.608606

[21237 rows x 4 columns]

Podemos considerar que os resultados parecem promissores, já que a coluna de avaliação prevista está bem próxima da avaliação real

Aqui estamos criando uma função que pega o CIE de uma escola, o CNPJ de uma fornecedora é aplica no modelo. O resultado é a nota prevista que aquela escola daria para o fornecedor.

In [ ]:
def fazer_previsao(CIE_escola, CNPJ_fornecedor):
    previsao = model.predict(CIE_escola, CNPJ_fornecedor)
    return previsao.est


fazer_previsao('122245','13.533.610/0001-00') # Previsão de uma escola que ainda não teve entrega com certa fornecedora
# Pelo resultado podemos analisar que a nota que ela daria para essa fornecedora seria essa abaixo.

3.8768484912358434

Nesse trecho estamos exportando o modelo utilizando joblib. Esse modelo é o que será utilizado na API.

In [ ]:
modelo = model
#Enviando o modelo para o joblib
joblib.dump(modelo, 'modelo_recomendacao.joblib')

['modelo_recomendacao.joblib']

Aqui criamos uma função que pega o CIE de uma escola e faz a previsão daquela escola com todos fornecedores disponíveis e retornar a nota prevista o nome daquele fornecedor que teria a maior avaliação.

In [ ]:
def melhor_fornecedora(CIE):
    df_fornecedoras_unicas = df_agrupado.drop_duplicates(subset='CNPJ')

    maior_valor = float('-inf')  # Inicializa com o menor valor possível
    nome_fornecedora_maior_valor = None

    for fornecedora in df_fornecedoras_unicas['CNPJ']:
        resultado_previsao = fazer_previsao(CIE, fornecedora)

        if resultado_previsao > maior_valor:
            maior_valor = resultado_previsao
            nome_fornecedora_maior_valor = fornecedora

    return nome_fornecedora_maior_valor, maior_valor

nome_fornecedora, maior_valor = melhor_fornecedora('12')
print(f'Maior valor retornado pela previsão: {maior_valor} para a fornecedora {nome_fornecedora}')



Maior valor retornado pela previsão: 4.612529898437838 para a fornecedora 71.615.942/0001-22


Podemos observar que o resultado para a escola de CIE 12 daria nota 4.61 para a fornecedora de CNPJ "71.615.942/0001-22". Sendo assim essa seria a ideal, já que essa nota foi comparada com todos fornecedores existentes e essa fornecedora teve maior nota.

In [ ]:
#df_agrupado.to_csv('fornecedoras.csv', index=False)
#files.download('fornecedoras.csv')